### ISSUE TITLE
Termdebug: add frames window and :Frames, similar to :Asm

### ISSUE THREAD

**Issue by GitMensch** [vim/vim#9235]
A new window for showing the current frames and highlight (similar to the assembler window) the current place would be nice :-) Likely more often used than the assembly window (which is definitely cool to have, too).

I _assume_ that this would share much of the dissasembly code but using `backtrace` command instead of `dissasembly`.

**Comment by lacygoill** [comment]
This is just a suggestion as I'm not familiar enough neither with debugging, nor with this new window, but if you think you end up with too many windows, you might be interested in a popup window, because the latter can be displayed over the others, and can be toggled on and off (see [`:help popup_hide()`](https://vimhelp.org/popup.txt.html#popup_hide%28%29) and [`:help popup_show()`](https://vimhelp.org/popup.txt.html#popup_show%28%29)).  Although that would only be OK if you don't intend to focus it; i.e. you just want to be able to read some info.

You might also try to set every frame as an entry in the quickfix list (or the location list); this way, you could leverage all the builtin commands around the quickfix window like `:cnext`, `:cfirst` ... and friends.

**Comment by lacygoill** [comment]
> You might also try to set every frame as an entry in the quickfix list (or the location list); this way, you could leverage all the builtin commands around the quickfix window like :cnext, :cfirst ... and friends.

On that front, you might be interested in [`:help getqflist-examples`](https://vimhelp.org/quickfix.txt.html#getqflist-examples), [`:help setqflist-examples`](https://vimhelp.org/quickfix.txt.html#setqflist-examples), and [`:help quickfix-window-function`](https://vimhelp.org/quickfix.txt.html#quickfix-window-function) (a concrete example is given at the end of the page).

**Comment by GitMensch** [comment]
Popups would be space-wise interesting, but in one of my personal use cases not usable - #3208. (Or did I misinterpret that and a popup is totally different from balloons?)

I'm not very familiar with the quickfix support - but that is __very likely the best way to handle `:Frames`__ with the builtin commands. Thanks for pointing this out and for adding the help pointers.

**Comment by lacygoill** [comment]
> (Or did I misinterpret that and a popup is totally different from balloons?)

A popup is different than a balloon.  See this [Popup window gallery](https://github.com/vim/vim/wiki/Popup-window-gallery) page on the wiki for a few screenshots and get and idea of what you could achieve.

**Comment by GitMensch** [comment]
@brammool: Are you ok with a Frames Window? You know, just asking before I work more on that...

@ all: With the quickfix list being "likely doable": would we still want an extra command for switching the frames #9234 or should this be dropped because the quickfix list will be usable for the navigation and "enough"?

**Comment by GitMensch** [comment]
@lacygoill after a bit of hacking the quickfix list I see that the commands are nice - but the list isn't usable (or I've missed something).
Reason: We don't want to **ask vim** go to the _file_ on the given _line_ (which is possible to parse from the response of `-stack-list-frames`), but we want to **ask gdb** to go to the frame (and from the resulting gdb mi information **lead vim** to go there. Otherwise we'd be in a frame and all `:Evaluate` will still go to the original frame.

Also, as the stack frame may be always open a `:Frame 5` command would be useful in any case - no need to switch the window.

Any opinions and/or thoughts?

BTW: is there a reasonable syntax for a stacktrace or even the default GDB backtrace format for the Frames window?

**Comment by GitMensch** [comment]
Yes, the output of where, updated on each "stop" (so step, brreakpoint reached, ...).
I have navigation through frames implemented as `:Up`, `:Down` (both using cli commands as the extension currently doesn't keep track of the active frame) and `:Frame N`.

I _would like to_ "somehow" use the frames-window for navigation, but I don't know how to achieve this (it is just a buffer with text), and my try with quickfix list ended in "_gdb_ must select the frame, not _vim_ the file".

**Comment by brammool** [comment]
> Yes, the output of where, updated on each "stop" (so step, brreakpoint
> reached, ...).  I have navigation through frames implemented as `:Up`,
> `:Down` (both using cli commands as the extension currently doesn't
> keep track of the active frame) and `:Frame N`.
> 
> I _would like to_ "somehow" use the frames-window for navigation, but
> I don't know how to achieve this (it is just a buffer with text), and
> my try with quickfix list ended in "_gdb_ must select the frame, not
> _vim_ the file".

Using a quickfix window doesn't seem like the right solution, since it
works quite differently.  You can most likely use buffer-local mappings.
E.g. to handle a mouse click or pressing Enter on a certain line
(relating to a certain frame).  Proabaly using 'buftype' "nofile".

Sending the frame command to gdb should then result in Vim displaying
the code there.  That should already work.

-- 
To be rich is not the end, but only a change of worries.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by GitMensch** [comment]
"possible in general" != possible for me, yet;
but I try to polish the current "Frames" window (view only, but always up to date) and frame related commands, then do a PR.

I hope you can then add the buffer-local mappings (the features will be in as functions already) and, if I haven't done that already marking of the current frame.

**Comment by ubaldot** [comment]
> "possible in general" != possible for me, yet; but I try to polish the current "Frames" window (view only, but always up to date) and frame related commands, then do a PR.
> 
> I hope you can then add the buffer-local mappings (the features will be in as functions already) and, if I haven't done that already marking of the current frame.

Could you please point to the PR you made (if any)?

**Comment by errael** [comment]
> Could you please point to the PR you made (if any)?

Isn't that #10393?

There's some stuff mentioned in here and in the open PR. In here, the idea from @lacygoill  of presenting as a popup and toggling show/hide is interesting. And in the PR, there's comments from Bram about future direction of `termdebug`

**Comment by ubaldot** [comment]
Thanks! I haven't seen that, but it has been an interesting read. The idea of having popups is also very nice actually. The tricky part may be decided who should go to ordinary windows and who should go to popup. My first guess would be to use popups for whatever features the user could interact with (e.g. frame popup or breakpoint list popup, the user can hit enter on a popup item to perform an action).

**Comment by errael** [comment]
> Thanks! I haven't seen that, but it has been an interesting read. The idea of having popups is also very nice actually. The tricky part may be decided who should go to ordinary windows and who should go to popup.

Why not let the user decide? The choice could be provided in both the config file and while running. Note that an existing buffer can be put into a popup. For example, there could be command(s) provided by `termdebug`, say `Popup({buf})` or maybe `PopupWin({win})` and `PopupBuf({buf})`.

Certain windows/buffers could be excluded, and ...

> My first guess would be to use popups for whatever features the user could interact with (e.g. frame popup or breakpoint list popup, the user can hit enter on a popup item to perform an action)
